Eth1 connections

michaelneuder opened the pull request:
**What type of PR is this?**
 Feature

**What does this PR do? Why is it needed?**
This implements an API to get the info about current ETH1 connections. It was a feature request opened in https://github.com/prysmaticlabs/prysm/issues/9972. 

**Which issues(s) does this PR fix?**
Fixes https://github.com/prysmaticlabs/prysm/issues/9972. 

**Other notes for review**
1. I added functions to the `ChainInfoFetcher` interface rather than creating a new interface in the powchain package.
2. For the `beacon-chain/powchain/service_test.go`, I was only able to unit test `CurrentETH1Endpoint()` and `ETH1Endpoints()`. I couldn't find a test that calls  `dialETH1Nodes`. Since the functions are really quite simple, hopefully living without the coverage is OK. But let me know if there is something I am missing.
3. For `beacon-chain/rpc/prysm/v1alpha1/node/service_test.go` I needed a mock that implemented the `powchain.ChainInfoFetcher` interface. I added it to the `testutil` package in `beacon-chain/rpc/testutil/mock_powchain_info_fetcher.go`. The mock just returns the fields it was constructed with or with the zero values.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/prysmaticlabs/prysm?pullRequest=10073) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/prysmaticlabs/prysm?pullRequest=10073) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: Michael Neuder<br/>:x: michaelneuder<br/><hr/>**Michael Neuder** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/prysmaticlabs/prysm?pullRequest=10073) it.</sub>

rauljordan commented:
Awesome work @michaelneuder this works as expected! I hope you don't mind I pushed a few commits to generate the protos and some Bazel fixes. I have just tested this out locally and it works as expected:

```
curl http://localhost:3500/eth/v1alpha1/node/eth1/connections
{"currentAddress":"http://localhost:8545","currentConnectionError":"Post \"http://localhost:8545\": dial tcp [::1]:8545: connect: connection refused","addresses":["http://localhost:8545"],"connectionErrors":["Post \"http://localhost:8545\": dial tcp [::1]:8545: connect: connection refused"]}%  
```
I even found out my eth1 node was down due to your feature, so thanks a lot :)

rauljordan commented:
For some reason our CLA assistant wants a second Michael to sign the agreement. Ill see if I can skip that, because it might have been from a commit on another computer you had

michaelneuder commented:
cool! thanks for the review!! I can't tell what I need to do next to merge the branch. It seems like some of the checks are failing?

rauljordan commented:
hi @michaelneuder I see some failures in building due to
https://github.com/prysmaticlabs/prysm/runs/4794010970?check_suite_focus=true

Once the PR is green, ill make an exception for the CLA assistant and merge

michaelneuder commented:
Ok I made a few more changes in regards to the errors we were looking at. I think I messed up the git state a bit, but it seems like all the changes are here. A few notes:

1. I added `GetETH1ConnectionStatus` to `testing/mock/node_service_mock.go`. This fixes one of the interface issues. 
2. I added the mock interfaces for the powchain side to `beacon-chain/powchain/testing/mock_powchain.go`. This fixed the other interface issue, and also allowed me to remove my mock fetcher, because I can just use mockPOW.POWChain struct in my `beacon-chain/rpc/prysm/v1alpha1/node/server_test.go`. 

Hoping this fixes the checks that we were erroring out on.

nisdas commented on beacon-chain/powchain/service.go:
```diff
@@ -324,6 +328,37 @@ func (s *Service) IsConnectedToETH1() bool {
+
+// CurrentETH1ConnectionError returns the error (if any) of the current connection.
+func (s *Service) CurrentETH1ConnectionError() error {
+	_, _, err := s.dialETH1Nodes(s.cfg.currHttpEndpoint)
```
The clients should be closed after we dial them otherwise there is a resource leak and the connection is
kept active.

michaelneuder replied:
sorry, totally missed this. thanks for addressing!

nisdas commented on beacon-chain/powchain/service.go:
```diff
@@ -324,6 +328,37 @@ func (s *Service) IsConnectedToETH1() bool {
+func (s *Service) ETH1ConnectionErrors() []error {
+	var errs []error
+	for _, ep := range s.cfg.httpEndpoints {
+		_, _, err := s.dialETH1Nodes(ep)
```
Same thing applies here

nisdas reviewed (commented):
It seems I was too late to review this before it was merged in, but there is a resource leak here with regards to dialing the eth1 json-rpc server.

prestonvanloon commented:
@michaelneuder Could you kindly address the feedback from @nisdas? 
We'll need to fix those concerns or revert these changes prior to next release. 
Thanks!

michaelneuder commented:
Sorry missed these comments!
